# Imports

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

# Loading the data

In [2]:
df = pd.read_csv('../data/processed/preprocessed-data.csv', index_col=0, encoding = 'ISO-8859-1')
df

location      age        isbn  rating  \
user_id                                                               
3329            grantsville, utah, usa  34.7439  0440234743       8   
7346        sunnyvale, california, usa  49.0000  0440234743       9   
7352               houston, texas, usa  53.0000  0440234743       8   
9419             somewhere, texas, usa  34.7439  0440234743       5   
11224        tumwater, washington, usa  51.0000  0440234743       6   
...                                ...      ...         ...     ...   
223566      kansas city, missouri, usa  29.0000  1880418568       8   
242083       fairview, new mexico, usa  24.0000  1880418568       7   
245604           napa, california, usa  43.0000  1880418568       9   
254995      san diego, california, usa  39.0000  1880418568       7   
259259   woodruff, south carolina, usa  30.0000  1880418568      10   

                                           book_title   book_author  \
user_id                                                               
3329                                    The Testament  John Grisham   
7346                                    The Testament  John Grisham   
7352                                    The Testament  John Grisham   
9419                                    The Testament  John Grisham   
11224                                   The Testament  John Grisham   
...                                               ...           ...   
223566   Wolves of the Calla (The Dark Tower, Book 5)  Stephen King   
242083   Wolves of the Calla (The Dark Tower, Book 5)  Stephen King   
245604   Wolves of the Calla (The Dark Tower, Book 5)  Stephen King   
254995   Wolves of the Calla (The Dark Tower, Book 5)  Stephen King   
259259   Wolves of the Calla (The Dark Tower, Book 5)  Stephen King   

         year_of_publication                 publisher  \
user_id                                                  
3329                  1999.0                      Dell   
7346                  1999.0                      Dell   
7352                  1999.0                      Dell   
9419                  1999.0                      Dell   
11224                 1999.0                      Dell   
...                      ...                       ...   
223566                2003.0  Donald M. Grant/Scribner   
242083                2003.0  Donald M. Grant/Scribner   
245604                2003.0  Donald M. Grant/Scribner   
254995                2003.0  Donald M. Grant/Scribner   
259259                2003.0  Donald M. Grant/Scribner   

                                                     img_s  \
user_id                                                      
3329     http://images.amazon.com/images/P/0440234743.0...   
7346     http://images.amazon.com/images/P/0440234743.0...   
7352     http://images.amazon.com/images/P/0440234743.0...   
9419     http://images.amazon.com/images/P/0440234743.0...   
11224    http://images.amazon.com/images/P/0440234743.0...   
...                                                    ...   
223566   http://images.amazon.com/images/P/1880418568.0...   
242083   http://images.amazon.com/images/P/1880418568.0...   
245604   http://images.amazon.com/images/P/1880418568.0...   
254995   http://images.amazon.com/images/P/1880418568.0...   
259259   http://images.amazon.com/images/P/1880418568.0...   

                                                     img_m  \
user_id                                                      
3329     http://images.amazon.com/images/P/0440234743.0...   
7346     http://images.amazon.com/images/P/0440234743.0...   
7352     http://images.amazon.com/images/P/0440234743.0...   
9419     http://images.amazon.com/images/P/0440234743.0...   
11224    http://images.amazon.com/images/P/0440234743.0...   
...                                                    ...   
223566   http://images.amazon.com/images/P/1880418568.0...   
242083   http://images.amazon.com/images/P/1880418568.0...   


In [4]:
users   = pd.read_csv('../data/raw/BX-Users.csv', delimiter = ";",encoding = 'ISO-8859-1', names=['user_id','location','age'], skiprows=1)
books   = pd.read_csv('../data/raw/BX_Books.csv', delimiter = ";",encoding = 'ISO-8859-1', names=['isbn','book_title','book_author','year_of_publication', 'publisher', 'img_s', 'img_m', 'img_l'], skiprows=1)
ratings = pd.read_csv('../data/raw/BX-Book-Ratings.csv', delimiter = ";",encoding = 'ISO-8859-1',names=['user_id','isbn','rating'], skiprows=1)


# Visualizing and Exploring

In [5]:
display(df.head())
print(df.shape)

user_id                   location      age        isbn  rating  \
0        2  stockton, california, usa  18.0000  0195153448       0   
1        8   timmins, ontario, canada  34.7439  0002005018       5   
2    11400    ottawa, ontario, canada  49.0000  0002005018       0   
3    11676              n/a, n/a, n/a  34.7439  0002005018       8   
4    41385   sudbury, ontario, canada  34.7439  0002005018       0   

            book_title           book_author  year_of_publication  \
0  Classical Mythology    Mark P. O. Morford               2002.0   
1         Clara Callan  Richard Bruce Wright               2001.0   
2         Clara Callan  Richard Bruce Wright               2001.0   
3         Clara Callan  Richard Bruce Wright               2001.0   
4         Clara Callan  Richard Bruce Wright               2001.0   

                 publisher                                              img_s  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
3    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
4    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                               img_m  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                               img_l  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                             Summary Language  \
0  Provides an introduction to classical myths pl...       en   
1  In a small town in Canada, Clara Callan reluct...       en   
2  In a small town in Canada, Clara Callan reluct...       en   
3  In a small town in Canada, Clara Callan reluct...       en   
4  In a small town in Canada, Clara Callan reluct...       en   

             Category      city       state country  
0  ['Social Science']  stockton  california     usa  
1       ['Actresses']   timmins     ontario  canada  
2       ['Actresses']    ottawa     ontario  canada  
3       ['Actresses']       NaN         NaN     NaN  
4       ['Actresses']   sudbury     ontario  canada

(1031175, 18)


In [6]:
users.head()

user_id                            location   age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [7]:
display(books.head())
print(books.shape)

isbn                                         book_title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            book_author  year_of_publication                publisher  \
0    Mark P. O. Morford                 2002  Oxford University Press   
1  Richard Bruce Wright                 2001    HarperFlamingo Canada   
2          Carlo D'Este                 1991          HarperPerennial   
3      Gina Bari Kolata                 1999     Farrar Straus Giroux   
4       E. J. W. Barber                 1999   W. W. Norton & Company   

                                               img_s  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                               img_m  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                               img_l  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

(271379, 8)


In [8]:
display(ratings.head())
print(ratings.shape)

user_id        isbn  rating
0   276725  034545104X       0
1   276726  0155061224       5
2   276727  0446520802       0
3   276729  052165615X       3
4   276729  0521795028       6

(1149780, 3)


Looking at the shape of the two data frames `df` (`1,031,175 rows`), which is the preprocessed one, and `ratings` (`1,149,780 rows`) we see a discrepancy of `118,605` ratings

In [9]:
combined = pd.merge(ratings, users, on='user_id', how='inner')

display(combined.head())
print(combined.shape)

user_id        isbn  rating                       location   age
0   276725  034545104X       0              tyler, texas, usa   NaN
1   276726  0155061224       5       seattle, washington, usa   NaN
2   276727  0446520802       0  h, new south wales, australia  16.0
3   276729  052165615X       3           rijeka, n/a, croatia  16.0
4   276729  0521795028       6           rijeka, n/a, croatia  16.0

(1149780, 5)


merging the original `ratings` dataframe on the `users` datafram with an inner join (i.e. keeping those rows where the key is in both dataframes), we see that the number of ratings does not change. Thus, all users in the `ratings` dataframe are covered in the `users` dataframe.

In [10]:

display(combined.head())
print(combined.shape)

user_id        isbn  rating            book_title book_author  \
0   276725  034545104X       0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X       5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X       0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X       5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X       9  Flesh Tones: A Novel  M. J. Rose   

   year_of_publication         publisher  \
0                 2002  Ballantine Books   
1                 2002  Ballantine Books   
2                 2002  Ballantine Books   
3                 2002  Ballantine Books   
4                 2002  Ballantine Books   

                                               img_s  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                               img_m  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                               img_l  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

(1031175, 10)


applying the same for `ratings` and `books` dataframes we see how the discrepancy between `ratings` and the preprocessed dataframe `df` occur.

It seems that the `books` dataframe has some `ISBN`s missing. Thus, ratings where there is no match on the `books` dataframe are dropped.

This shows that the `df` dataframe from the `Preprocessed_data.csv` file is valid. For further exploration the `df` dataframe is used as it also has the three features `Summary`, `Language`, and `Category`.

# Preprocessing

to have a not so sparse dataframe we do want to only consider the following ratings:
- ratings where the rating is between 1 and 10 (only explicit ratings, no 0 rating)
- books that have been rated at least by 20 different users

In [11]:
#Function printing the number of unique users and books in the dataframe
def print_information(df, name):
    df_num_ratings =  df.shape[0]
    df_num_users = len(df['user_id'].unique())
    df_num_books = len(df['isbn'].unique())
    print(f'For {name}:\n\tNumber of unique users: {df_num_users}\n\tNumber of unique books: {df_num_books}\n\tNumber of ratings: {df_num_ratings}')

In [12]:
#Information about the dataframe before preprocessing
print_information(df, 'original data frame')


For original data frame:
	Number of unique users: 92107
	Number of unique books: 270170
	Number of ratings: 1031175


In [13]:
#only taking rating between 1 and 10
df_no_zeros = df.loc[df['rating'] > 0]
print_information(df_no_zeros, 'dataframe only books rated with 1 or higher (i.e. no 0 ratings)')
df_no_zeros['rating'].value_counts().sort_values()


For dataframe only books rated with 1 or higher (i.e. no 0 ratings):
	Number of unique users: 68092
	Number of unique books: 149842
	Number of ratings: 383852


1      1481
2      2375
3      5118
4      7617
6     31689
5     45355
9     60780
7     66404
10    71227
8     91806
Name: rating, dtype: int64

In [14]:
#only taking ratings of books where the book was rated at least 20 times
df_10_ratings_least = df_no_zeros.groupby('isbn').filter(lambda df: df.shape[0] > 9)
print_information(df_10_ratings_least, 'dataframe only books rated more than 10 times')

#only taking ratings of books where the book was rated at least 20 times
df_15_ratings_least = df_no_zeros.groupby('isbn').filter(lambda df: df.shape[0] > 14)
print_information(df_15_ratings_least, 'dataframe only books rated more than 15 times')

#only taking ratings of books where the book was rated at least 20 times
df_20_ratings_least = df_no_zeros.groupby('isbn').filter(lambda df: df.shape[0] > 19)
print_information(df_20_ratings_least, 'dataframe only books rated more than 20 times')


For dataframe only books rated more than 10 times:
	Number of unique users: 39365
	Number of unique books: 5444
	Number of ratings: 139296
For dataframe only books rated more than 15 times:
	Number of unique users: 34573
	Number of unique books: 3151
	Number of ratings: 112631
For dataframe only books rated more than 20 times:
	Number of unique users: 31383
	Number of unique books: 2129
	Number of ratings: 95580


# Training a Model

for the purpose of the recommendation system we want to train different models and see which one performs best.
The following models will be used
- SVD
- User Based Collaborative Filtering 
- Item Based Collaborative Filtering 

In [15]:
# Install the surprise package
%pip install scikit-surprise

Note: you may need to restart the kernel to use updated packages.


In [16]:
from surprise import SVD, NMF, KNNBasic, KNNWithMeans, Dataset, Reader, accuracy
from sklearn.model_selection import train_test_split
from surprise.model_selection import cross_validate


## Load the data into surprise input data format, split into training and testing
reader = Reader(rating_scale=(1, 10))
trainset, testset = train_test_split(df_10_ratings_least[['user_id','isbn','rating']], test_size=0.25, random_state=0)
trainset.index = range(len(trainset)) 
testset.index = range(len(testset))
trainset = Dataset.load_from_df(trainset, reader)
testset = testset.values.tolist()

# Define the models and their hyperparameters 
sim_option_item = {'name': 'cosine',
              'user_based': False  # compute similarities between items
               }
sim_option_user = {'name': 'pearson',
              'user_based': True  # compute similarities between users
               }
alg_list = ['SVD(n_factors=50)', 'NMF(n_factors=15)',
            'KNNBasic(k=20, sim=cosine, item-based)', 
            'KNNBasic(k=20, sim=pearson, user-based)']

i = 0

benchmark = []


# Iterate over selected algorithms based on 5-fold CV on training set
for algorithm in [SVD(n_factors=30), NMF(n_factors=15), KNNBasic(k=20, sim_options=sim_option_item), KNNBasic(k=20, sim_options=sim_option_user)]:
    # Perform cross validation
    results = cross_validate(algorithm, trainset, measures=['RMSE','MAE'], cv=5, verbose=False)
    
    # Get results & append algorithm name
    tmp = pd.DataFrame.from_dict(results).mean(axis=0)
    tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))
    benchmark.append(tmp)
    i+=1
pd.DataFrame(benchmark).set_index('Algorithm').sort_values('test_rmse')    

/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/1830921147.py:34: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


test_rmse  test_mae  fit_time  test_time
Algorithm                                                  
SVD(n_factors=50)    1.61473  1.250534   3.63672   0.184497

In [17]:
from surprise import SVD, NMF, KNNBasic, KNNWithMeans, Dataset, Reader, accuracy
from sklearn.model_selection import train_test_split
from surprise.model_selection import cross_validate


## Load the data into surprise input data format, split into training and testing
reader = Reader(rating_scale=(1, 10))
trainset, testset = train_test_split(df_10_ratings_least[['user_id','isbn','rating']], test_size=0.25, random_state=0)
trainset.index = range(len(trainset)) 
testset.index = range(len(testset))
trainset = Dataset.load_from_df(trainset, reader)
testset = testset.values.tolist()


benchmark = []
i = 0
alg_list = ['']


# Iterate over selected algorithms based on 5-fold CV on training set
for n in range(5,50,5):#, NMF(n_factors=15), KNNBasic(k=20, sim_options=sim_option_item), KNNBasic(k=20, sim_options=sim_option_user)]:
    # Perform cross validation
    alg_list.append(str(n))
    print(n)
    algorithm = SVD(n_factors=n)
    results = cross_validate(algorithm, trainset, measures=['RMSE','MAE'], cv=5, verbose=False)
    
    # Get results & append algorithm name
    tmp = pd.DataFrame.from_dict(results).mean(axis=0)
    print(f'Dimensions: {n}')
    display(tmp)

    tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))
    benchmark.append(tmp)
    i+=1
pd.DataFrame(benchmark).set_index('Algorithm').sort_values('test_rmse')

5
Dimensions: 5


test_rmse    1.610759
test_mae     1.247307
fit_time     1.662944
test_time    0.171707
dtype: float64

10


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 10


test_rmse    1.613273
test_mae     1.249570
fit_time     1.918559
test_time    0.150474
dtype: float64

15


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 15


test_rmse    1.613857
test_mae     1.249606
fit_time     2.382491
test_time    0.159978
dtype: float64

20


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 20


test_rmse    1.614546
test_mae     1.250449
fit_time     2.273441
test_time    0.149492
dtype: float64

25


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 25


test_rmse    1.615648
test_mae     1.251237
fit_time     2.570838
test_time    0.171872
dtype: float64

30


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 30


test_rmse    1.613880
test_mae     1.250181
fit_time     3.384465
test_time    0.173794
dtype: float64

35


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 35


test_rmse    1.615349
test_mae     1.251364
fit_time     3.139248
test_time    0.191417
dtype: float64

40


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 40


test_rmse    1.614680
test_mae     1.251038
fit_time     3.278906
test_time    0.174232
dtype: float64

45


/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


Dimensions: 45


test_rmse    1.616727
test_mae     1.252263
fit_time     3.518142
test_time    0.136461
dtype: float64

/var/folders/gl/fxpgq_v15qn2fvg77jff1p100000gn/T/ipykernel_24435/3339071703.py:33: FutureWarning: The series.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tmp = tmp.append(pd.Series([alg_list[i]], index=['Algorithm']))


test_rmse  test_mae  fit_time  test_time
Algorithm                                          
            1.610759  1.247307  1.662944   0.171707
5           1.613273  1.249570  1.918559   0.150474
10          1.613857  1.249606  2.382491   0.159978
25          1.613880  1.250181  3.384465   0.173794
15          1.614546  1.250449  2.273441   0.149492
35          1.614680  1.251038  3.278906   0.174232
30          1.615349  1.251364  3.139248   0.191417
20          1.615648  1.251237  2.570838   0.171872
40          1.616727  1.252263  3.518142   0.136461

# Content Based Filtering

In [25]:
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

In [38]:
books

isbn                                         book_title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271374  0440400988                         There's a Bat in Bunk Five   
271375  0525447644                            From One to One Hundred   
271376  006008667X  Lily Dale : The True Story of the Town that Ta...   
271377  0192126040                        Republic (World's Classics)   
271378  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 book_author  year_of_publication  \
0         Mark P. O. Morford                 2002   
1       Richard Bruce Wright                 2001   
2               Carlo D'Este                 1991   
3           Gina Bari Kolata                 1999   
4            E. J. W. Barber                 1999   
...                      ...                  ...   
271374        Paula Danziger                 1988   
271375            Teri Sloat                 1991   
271376      Christine Wicker                 2004   
271377                 Plato                 1996   
271378   Christopher  Biffle                 2000   

                                               publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                                 W. W. Norton & Company   
...                                                  ...   
271374                   Random House Childrens Pub (Mm)   
271375                                      Dutton Books   
271376                                HarperSanFrancisco   
271377                           Oxford University Press   
271378  McGraw-Hill Humanities/Social Sciences/Languages   

                                                    img_s  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271374  http://images.amazon.com/images/P/0440400988.0...   
271375  http://images.amazon.com/images/P/0525447644.0...   
271376  http://images.amazon.com/images/P/006008667X.0...   
271377  http://images.amazon.com/images/P/0192126040.0...   
271378  http://images.amazon.com/images/P/0767409752.0...   

                                                    img_m  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271374  http://images.amazon.com/images/P/0440400988.0...   
271375  http://images.amazon.com/images/P/0525447644.0...   
271376  http://images.amazon.com/images/P/006008667X.0...   
271377  http://images.amazon.com/images/P/0192126040.0...   
271378  http://images.amazon.com/images/P/0767409752.0...   

                                                    img_l  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http:

In [50]:
summaries = df.loc[:,['isbn','Summary']]
summaries = summaries.groupby('isbn').first().reset_index()
summaries

books_w_summary = pd.merge(books, summaries, on='isbn', how='left')
books_w_summary = books_w_summary.loc[:, ['isbn','Summary']]
#replace occurances of '9' in the summary with NaN
books_w_summary['Summary'] = books_w_summary.Summary.apply(lambda x: np.NaN if x == '9' else x)

In [54]:
books_w_summary = books_w_summary[books_w_summary['Summary'].notna()]
books_w_summary

isbn                                            Summary
0       0195153448  Provides an introduction to classical myths pl...
1       0002005018  In a small town in Canada, Clara Callan reluct...
2       0060973129  Here, for the first time in paperback, is an o...
3       0374157065  Describes the great flu epidemic of 1918, an o...
4       0393045218  A look at the incredibly well-preserved ancien...
...            ...                                                ...
271373  0449906736  Drawing on their reporting experience from mor...
271374  0440400988  On her own for the first time, fourteen-year-o...
271375  0525447644  Illustrations of people and animals introduce ...
271376  006008667X  An award-winning and wry journalist captures t...
271377  0192126040  The central work of one of the West&#39;s grea...

[141967 rows x 2 columns]

In [57]:
vectorizer = TfidfVectorizer(lowercase=True, stop_words='english', use_idf=True)
X = vectorizer.fit_transform(books_w_summary['Summary'])

frozenset({'a',
           'about',
           'above',
           'across',
           'after',
           'afterwards',
           'again',
           'against',
           'all',
           'almost',
           'alone',
           'along',
           'already',
           'also',
           'although',
           'always',
           'am',
           'among',
           'amongst',
           'amoungst',
           'amount',
           'an',
           'and',
           'another',
           'any',
           'anyhow',
           'anyone',
           'anything',
           'anyway',
           'anywhere',
           'are',
           'around',
           'as',
           'at',
           'back',
           'be',
           'became',
           'because',
           'become',
           'becomes',
           'becoming',
           'been',
           'before',
           'beforehand',
           'behind',
           'being',
           'below',
           'beside',
           'besides'

In [65]:
X.shape

(141967, 110449)

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
X_cos_matrix = cosine_similarity(X,X)

In [ ]:
indices = pd.Series(books_w_summary.index, index=books_w_summary['isbn']).drop_duplicates()

def get_recommendations(isbn, cosine_similarity, indices):
    #Get the index of the book that matches the ISBN
    index = indices[isbn]
    #Get the pairwise similarity scores
    sim_scores = list(enumerate(cosine_similarity[index]))

    #Sort the books based on the similarity scores (it's the second element, thus x[1])
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    #Get the score for the 10 most similar movies
    sim_scores = sim_scores[1:11]
